src/feature/hs/hs_ob.c

```diff
@@ -0,0 +1,182 @@
+#define HS_OB_PRIVATE
+
+#include "lib/confmgt/confmgt.h"
+#include "lib/encoding/confline.h"
+
+#include "hs_ob.h"
+
+/* Options config magic number. */
+#define OB_OPTIONS_MAGIC 0x631DE7EA
+
+/* Helper macros. */
+#define VAR(varname, conftype, member, initvalue)                          \
+  CONFIG_VAR_ETYPE(ob_options_t, varname, conftype, member, 0, initvalue)
+#define V(member,conftype,initvalue)        \
+  VAR(#member, conftype, member, initvalue)
+
+/* Dummy instance of ob_options_t, used for type-checking its members with
+ * CONF_CHECK_VAR_TYPE. */
+DUMMY_TYPECHECK_INSTANCE(ob_options_t);
+
+/* Array of variables for the config file options. */
+static const config_var_t config_vars[] = {
+  V(MasterOnionAddress, LINELIST, NULL),
+
+  END_OF_CONFIG_VARS
+};
+
+/* "Extra" variable in the state that receives lines we can't parse. This
+ * lets us preserve options from versions of Tor newer than us. */
+static const struct_member_t config_extra_vars = {
+  .name = "__extra",
+  .type = CONFIG_TYPE_LINELIST,
+  .offset = offsetof(ob_options_t, ExtraLines),
+};
+
+/* Configuration format of ob_options_t. */
+static const config_format_t config_format = {
+  .size = sizeof(ob_options_t),
+  .magic = {
+     "ob_options_t",
+     OB_OPTIONS_MAGIC,
+     offsetof(ob_options_t, magic_),
+  },
+  .vars = config_vars,
+  .extra = &config_extra_vars,
+};
+
+/* Global configuration manager for the config file. */
+static config_mgr_t *config_options_mgr = NULL;
+
+/* Return the configuration manager for the config file. */
+static const config_mgr_t *
+get_config_options_mgr(void)
+{
+  if (PREDICT_UNLIKELY(config_options_mgr == NULL)) {
+    config_options_mgr = config_mgr_new(&config_format);
+    config_mgr_freeze(config_options_mgr);
+  }
+  return config_options_mgr;
+}
+
+#define ob_option_free(val) \
+  FREE_AND_NULL(ob_options_t, ob_option_free_, (val))
+
+/** Helper: Free a config options object. */
+static void
+ob_option_free_(ob_options_t *opts)
+{
+  if (opts == NULL) {
+    return;
+  }
+  config_free(get_config_options_mgr(), opts);
+}
+
+/** Return an allocated config options object. */
+static ob_options_t *
+ob_option_new(void)
+{
+  ob_options_t *opts = config_new(get_config_options_mgr());
+  config_init(get_config_options_mgr(), opts);
+  return opts;
+}
+
+/** Parse the given ob options in opts and set the service config object
+ * accordingly.
+ *
+ * Return 1 on success else 0. */
+static int
+ob_option_parse(hs_service_config_t *config, const ob_options_t *opts)
+{
+  int ret = 0;
+  config_line_t *line;
+
+  tor_assert(config);
+  tor_assert(opts);
+
+  for (line = opts->MasterOnionAddress; line; line = line->next) {
+    /* Allocate config list if need be. */
+    if (!config->ob_master_pubkeys) {
+      config->ob_master_pubkeys = smartlist_new();
+    }
+    uint8_t version = 0;
+    ed25519_public_key_t *pubkey = tor_malloc_zero(sizeof(*pubkey));
+
+    if (hs_parse_address(line->value, pubkey, NULL, &version) < 0) {
```

Comment:
I get an assertion failure here:
```
[err] Bug: Tor 0.4.3.0-alpha-dev (git-77ef50d86b646f63): Assertion global_options failed in get_options_mutable at . Stack trace: (on Tor 0.4.3.0-alpha-dev 77ef50d86b646f63)
[err] Bug:     tor(log_backtrace_impl+0x56) [0x564e089aa2c6] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(tor_assertion_failed_+0x148) [0x564e089a5408] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(get_options_mutable+0x62) [0x564e0891b302] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(escaped_safe_str+0x1d) [0x564e0891b7cd] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(hs_parse_address+0xf8) [0x564e088a4748] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(hs_ob_parse_config_file+0x193) [0x564e0893ba83] (on Tor 0.4.3.0-alpha-dev
[err] Bug:     tor(hs_config_service_all+0xd6c) [0x564e0893a33c] (on Tor 0.4.3.0-alpha-dev 77ef50d86b646f63)
```

We seem to be trying to do this too early and it's invoking get_options when its not yet ready.

I dirtily fixed this by removing the calls to escaped log, but we need to do something better here.

Reply:
We'll have to do a special `hs_parse_address()` that returns an error string instead of directly logging. So then, we can use the `options_t` that `hs_config_service_all()` has accessed to.